Use composition for multiagents

Tortar opened the pull request:
I didn't know about this feature but it is sweet in this case I think :-)

Datseris reviewed (changes requested):
I didn't know about this either... But, If two people that develop such a complex library don't know something about Julia, what does this say for the average user...? I think the average user would have no clue what this piece of code even means. This seems a cool feature, but probably should be highlighted as an advanced illustration. My suggestion would be to keep the the tutorial as-is and _add_ another sentence afterwards that says something "actually, julia even allows for composition. You can use this character (with unicode \name and tab) and then compose functions". Or something like that?

Tortar commented:
yeah, I had the same feeling, let's go with suggesting it after the constructor calls

Datseris commented:
I've improved the tutorial a bit. @Tortar I noticed that the two approaches aren't interchangeable per se:
```
julia> polit_constr = constructor(MultiSchelling, Politician)
#4 (generic function with 1 method)

julia> MultiSchelling ∘ Schelling
MultiSchelling ∘ Schelling
```

I assume this doens't have any impact on performance? To finish this do you want to replace the syntax in the docstring of `@multiagent` with the compsosition unicode?

Tortar commented:
Both return a function so they are interchangeable for our scopes and the composition shouldn't change any perf characteristics.

Replaced the composition in multiagent, I also changed a bit the constructor stuff in the tutorial, let me know what do you think. In general to construct single agents we have three options:

- Multi(Agent(...))
- (Multi∘Agent)(...)
- constructor(Multi, Agent)(...)

I mean, in a sense, now that I know about the ∘ operator I would completely drop constructor(Multi, Agent)(...) and I would explain the logic more than anything else, e.g. if we woudn't have the composition operator we would need a closure to build the Multi(Agent(...)) syntax

Datseris commented on examples/event_rock_paper_scissors.jl:
```diff
@@ -191,7 +191,7 @@ const alltypes = (Rock, Paper, Scissors)
 for p in positions(model)
     type = rand(abmrng(model), alltypes)
-    add_agent!(p, constructor(RPS, type), model)
+    add_agent!(p, RPS∘type, model)
```
```suggestion
    add_agent!(p, RPS ∘ type, model)
```

Datseris commented on examples/event_rock_paper_scissors.jl:
```diff
@@ -214,7 +214,7 @@ function initialize_rps(; n = 100, nx = n, ny = n, seed = 42)
     for p in positions(model)
         type = rand(abmrng(model), alltypes)
-        add_agent!(p, constructor(RPS, type), model)
+        add_agent!(p, RPS∘type, model)
```
```suggestion
        add_agent!(p, RPS ∘ type, model)
```

Datseris commented on examples/rabbit_fox_hawk.jl:
```diff
@@ -149,17 +149,17 @@ function initialize_model(
     for _ in 1:n_rabbits
         pos = random_walkable(model, model.landfinder)
-        agent = Animal(Rabbit(model, random_position(model), v0, rand(abmrng(model), Δe_grass:2Δe_grass)))
+        agent = (Animal∘Rabbit)(model, pos, v0, rand(abmrng(model), Δe_grass:2Δe_grass))
```
```suggestion
        agent = (Animal ∘ Rabbit)(model, pos, v0, rand(abmrng(model), Δe_grass:2Δe_grass))
```

Datseris commented on examples/rabbit_fox_hawk.jl:
```diff
@@ -149,17 +149,17 @@ function initialize_model(
     for _ in 1:n_foxes
         pos = random_walkable(model, model.landfinder)
-        agent = Animal(Fox(model, random_position(model), v0, rand(abmrng(model), Δe_rabbit:2Δe_rabbit)))
+        agent = (Animal∘Fox)(model, pos, v0, rand(abmrng(model), Δe_rabbit:2Δe_rabbit))
```
```suggestion
        agent = (Animal ∘ Fox)(model, pos, v0, rand(abmrng(model), Δe_rabbit:2Δe_rabbit))
```

Datseris commented on examples/rabbit_fox_hawk.jl:
```diff
@@ -149,17 +149,17 @@ function initialize_model(
     for _ in 1:n_hawks
         pos = random_walkable(model, model.airfinder)
-        agent = Animal(Hawk(model, random_position(model), v0, rand(abmrng(model), Δe_rabbit:2Δe_rabbit)))
+        agent = (Animal∘Hawk)(model, pos, v0, rand(abmrng(model), Δe_rabbit:2Δe_rabbit))
```
```suggestion
        agent = (Animal ∘ Hawk)(model, pos, v0, rand(abmrng(model), Δe_rabbit:2Δe_rabbit))
```

Datseris commented on src/core/agents.jl:
```diff
@@ -287,10 +290,10 @@ end
-hawk_2 = constructor(Animal, Hawk)(; id = 2, pos = (1, 2), ground_speed = 2.3, flight_speed = 2)
-wolf_1 = constructor(Animal, Wolf)(3, (2, 2), 2.0, 3.0, :black)
-wolf_2 = constructor(Animal, Wolf)(; id = 4, pos = (2, 1), ground_speed = 2.0, fur_color = :white)
+hawk_1 = (Animal∘Hawk)(1, (1, 1), 1.0, 2.0, 3)
+hawk_2 = (Animal∘Hawk)(; id = 2, pos = (1, 2), ground_speed = 2.3, flight_speed = 2)
+wolf_1 = (Animal∘Wolf)(3, (2, 2), 2.0, 3.0, :black)
+wolf_2 = (Animal∘Wolf)(; id = 4, pos = (2, 1), ground_speed = 2.0, fur_color = :white)
```
```suggestion
hawk_1 = (Animal ∘ Hawk)(1, (1, 1), 1.0, 2.0, 3)
hawk_2 = (Animal ∘ Hawk)(; id = 2, pos = (1, 2), ground_speed = 2.3, flight_speed = 2)
wolf_1 = (Animal ∘ Wolf)(3, (2, 2), 2.0, 3.0, :black)
wolf_2 = (Animal ∘ Wolf)(; id = 4, pos = (2, 1), ground_speed = 2.0, fur_color = :white)
```
